chore(blog): new post about asyncapi activities in february (#200)
asyncapi/website: pages/blog/february-2021-at-asyncapi.md

Is an "open governance model" the same as a "charter"? No. Charter is a boring legal document that describes many rules that need to be followed when being under a foundation. One of the things the charter tackles is the general rules on how the project will operate and be governed. 

I think the most important thing to write here is that the AsyncAPI charter is a reality. It finally arrived and is ready for review. In the special blog post, we also explained what governance model we envision for the AsyncAPI Initiative in more human-friendly words. This content was released and communicated on Tuesday, 9th of March, and we will wait for your feedback until the 23rd of March. It looks like we will kick off April under the umbrella of the foundation :crossed_fingers:

Please take your time to:
- Read [Finding a Good Open Governance Model for AsyncAPI](/blog/governance-motivation) to understand our motivation
- Check out [this](https://github.com/asyncapi/.github/pull/37) pull request with the charter

## Code generators activities

We had some significant traffic in the area of code generation templates in February.

### PHP

> [Emiliano Zublena](https://twitter.com/emilianozublena) joined the AsyncAPI community big time by starting with donating [a new template for PHP](https://github.com/asyncapi/asyncapi-php-template). It is not yet released under **@asyncapi** scope on npm, but you can already play with it by using the AsyncAPI Generator with a direct GitHub link like:

Reviewer: Awesome work, thanks guys!

I know i might be late on this, I'm not active at all on Twitter, it might be a better idea to link either my github acc or linkedin acc :)

Author: @emilianozublena  please open PR with the fix and I merge

```bash
ag https://bit.ly/asyncapi https://github.com/asyncapi/asyncapi-php-template -o output`
```

### Go

> We merged the initial pull request to the [Go template](https://github.com/asyncapi/go-template). The initial contributor was not able to continue working on the template, but the foundation was there. [Emiliano Zublena](https://twitter.com/emilianozublena) and [Takumi Sueda](https://github.com/puhitaku) will try to help to drive forward template development. This template is not yet released as we need to get some feedback from the community first. Give it a try with:

Reviewer: Same as above :)

---
title: February 2021 at AsyncAPI

> date: 2021-02-11T06:00:00+01:00

Reviewer 2: ```suggestion
date: 2021-03-15T06:00:00+01:00
```

Author: will change it for 16.03 later, now I didn't want to do it as it would not be rendered in the UI

Reviewer 2: This is not the behavior anymore :) It used to be with Hugo but now we're not filtering by date or anything because the website is statically generated on deployment.

Author: oh 😄 changing to 16 then

type: Communication
tags:
  - status
cover: /img/posts/february-2021-at-asyncapi-cover.webp
authors:
  - name: Lukasz Gornicki
    photo: /img/avatars/lpgornicki.webp
    link: https://twitter.com/derberq
    byline: AsyncAPI Maintainer and Community Guardian

> excerpt: We are getting close to joining a foundation. We started few new initiatives. We are busy. In other words, this is a good time to join us and drive things together

Reviewer 2: ```suggestion
excerpt: We are getting close to joining a foundation. We started a few new initiatives. We are busy. In other words, this is a good time to join us and drive things together.
```

featured: true
---

> Read [January 2021 at AsyncAPI](/blog/january-2021-at-asyncapi) for the update from January.

> I enjoy monthly status over weekly one. It is not that I'm just happy I have more time for other things. Content size did not change. I think people do have other things to do than follow AsyncAPI only :sweat_smile:. Life in open-source runs slower, at least when you look at it from the outside. I have an impression more people pay attention to the status when it is once a month. I wonder what your view is on that.

Reviewer 2: ```suggestion
I enjoy monthly status over weekly one. It is not that I'm just happy I have more time for other things. Content size did not change. I think people do have other things to do than following AsyncAPI only :sweat_smile: Life in open-source runs slower, at least when you look at it from the outside. I have an impression more people pay attention to the status when it is once a month. I wonder what your view is on that.
```

## Open governance model aka charter ready for the review

Is an "open governance model" the same as a "charter"? No. Charter is a boring legal document that describes many rules that need to be followed when being under a foundation. One of the things the charter tackles is the general rules on how the project will operate and be governed. 

> I think more important to write here is that the charter is here. It finally arrived and is ready for review. In the special blog post, we also explained what governance model we envision for the AsyncAPI Initiative in more human-friendly words. This content was released and communicated on Tuesday, 9th of March, and we will wait for your feedback until 23rd of March. It looks like we will kick off April under the umbrella of the foundation :crossed_fingers:

Reviewer 2: ```suggestion
I think the most important thing to write here is that the AsyncAPI charter is a reality. It finally arrived and is ready for review. In the special blog post, we also explained what governance model we envision for the AsyncAPI Initiative in more human-friendly words. This content was released and communicated on Tuesday, 9th of March, and we will wait for your feedback until the 23rd of March. It looks like we will kick off April under the umbrella of the foundation :crossed_fingers:
```

Please take your time to:
- Read [Finding a Good Open Governance Model for AsyncAPI](/blog/governance-motivation) to understand our motivation
- Check out [this](https://github.com/asyncapi/.github/pull/37) pull request with the charter

## Code generators activities

We had some significant traffic in the area of code generation templates in February.

### PHP

> [Emiliano Zublena](https://twitter.com/emilianozublena) joined the AsyncAPI community big time by starting with donating [a new template for PHP](https://github.com/asyncapi/asyncapi-php-template). It is not yet released under **@asyncapi** scope on npm, but you can already play with it by using the AsyncAPI Generator with direct GitHub link like:

Reviewer 2: ```suggestion
[Emiliano Zublena](https://twitter.com/emilianozublena) joined the AsyncAPI community big time by starting with donating [a new template for PHP](https://github.com/asyncapi/asyncapi-php-template). It is not yet released under **@asyncapi** scope on npm, but you can already play with it by using the AsyncAPI Generator with a direct GitHub link like:
```

```bash
ag https://bit.ly/asyncapi https://github.com/asyncapi/asyncapi-php-template -o output`
```

### Go

> We merged the initial pull request to the [Go template](https://github.com/asyncapi/go-template). The initial contributor was not able to continue working on the template, but the foundation is there. [Emiliano Zublena](https://twitter.com/emilianozublena) and [Takumi Sueda](https://github.com/puhitaku) will try to help to drive forward template development. This template is not yet released as we need to get some feedback from the community first. Give it a try with:

Reviewer 2: ```suggestion
We merged the initial pull request to the [Go template](https://github.com/asyncapi/go-template). The initial contributor was not able to continue working on the template, but the foundation was there. [Emiliano Zublena](https://twitter.com/emilianozublena) and [Takumi Sueda](https://github.com/puhitaku) will try to help to drive forward template development. This template is not yet released as we need to get some feedback from the community first. Give it a try with:
```

```bash
ag https://bit.ly/asyncapi https://github.com/asyncapi/go-template -o output
```

> Let us know what you think in the GitHub issues section. Thank you [Jacob Poston](https://github.com/jposton96a) for your initial hard work on the template!

Reviewer 2: ```suggestion
Let us know what you think in the GitHub issues section. Thank you, [Jacob Poston](https://github.com/jposton96a) for your initial hard work on the template!
```

### TypeScript and NATS

[Jonas Lagoni](https://github.com/jonaslagoni/) regularly extends and polishes (the TypeScript template for NATS](https://github.com/asyncapi/ts-nats-template). It reached 0.3 release and is the first template that is already using a new AsyncAPI SDK for data types generation. Please go and check it out.

## AsyncAPI CLI to rule them all

> We always wanted to have a single CLI to do all the things related to AsyncAPI. In other words, one CLI to validate, generate, edit, create and others. So far, we had CLI for the AsyncAPI generator, and recently one of the community members, [Jorge Aguiar Martín](https://twitter.com/jotamusik), referred us to his CLI for AsyncAPI validation.

Reviewer 2: ```suggestion
We always wanted to have a single CLI to do all the things related to AsyncAPI. In other words, one CLI to validate, generate, edit, create, and others. So far, we had a CLI for the AsyncAPI generator, and recently one of our community members, [Jorge Aguiar Martín](https://twitter.com/jotamusik), referred us to his CLI for AsyncAPI validation.
```

> We all agreed that instead of working separately on different CLIs, it is better to work together on one CLI that everyone will love. The project kicked off! Repository created, and we already discuss details on how the CLI interface should look like. [Join](https://github.com/asyncapi/cli/issues/1) the discussion.

Reviewer 2: ```suggestion
We all agreed that instead of working separately on different CLIs, it is better to work together on one CLI that everyone will love. The project kicked off! A repository has been created and we already discuss details on how the CLI interface should look like. [Join](https://github.com/asyncapi/cli/issues/1) the discussion.
```

## Event gateway

Sounds big? It is :smiley:

An excellent engineer joined AsyncAPI, [Sergio Moya](https://twitter.com/smoyac/status/1361289838075539461). We felt like we need to start something big that should be built independently from any vendor together with the AsyncAPI community. The project got kicked off, and it is the best time to join. Have a look at the dedicated repository and our plans for [the Everest? AsyncAPI Gate? or maybe Eventide?](https://github.com/asyncapi/event-gateway) :smiley:

Share your use cases. Please help us understand what you need. Sergio prepared an issue template that [helps with that](https://github.com/asyncapi/event-gateway/issues/new?assignees=&labels=use+case&template=use_case.md&title=%5BUSECASE%5D+).

## Domain model generation

> As mentioned in [the blog post about January](/blog/january-2021-at-asyncapi#next-major-feature-is-data-model-generation), February was all about data model generation that we wanted to use to speed up templates development for code generation. The library is [ready](https://github.com/asyncapi/generator-model-sdk), and we started trying it out. You can try it out too, and you do not even need our AsyncAPI Generator or AsyncAPI at all. It is a generic library and you only need JSON Schema Draft 7.

Reviewer 3: It supports both JSON Schema and AsyncAPI documents as inputs :)
```suggestion
As mentioned in [the blog post about January](/blog/january-2021-at-asyncapi#next-major-feature-is-data-model-generation), February was all about data model generation that we wanted to use to speed up templates development for code generation. The library is [ready](https://github.com/asyncapi/generator-model-sdk), and we started trying it out. You can try it out too, and you do not even need our AsyncAPI Generator. It is a generic library and you only need a JSON Schema Draft 7 or an AsyncAPI document.
```

```js
import { TypeScriptGenerator } from '@asyncapi/generator-model-sdk';

const DESCRIPTION_PRESET = {
  interface: {
    property({ property, content }) {
      const desc = property.getFromSchema('description');
      if (desc) {
        return `// ${desc}\n${content}`;
      }
      return content;
    }
  }
}

const options = {
  modelType: 'interface',
  presets: [DESCRIPTION_PRESET],
}

const generator = new TypeScriptGenerator(options);

const schema = {
  $id: "Address",
  type: "object",
  properties: {
    street_name:    { type: "string" },
    city:           { type: "string", description: "City description" },

>     state:          { type: "string" },

Reviewer 4: ```suggestion
```

    house_number:   { type: "number" },
    marriage:       { type: "boolean", description: "Status if marriage live in given house" },
    pet_names:      { type: "array", items: { type: "string" } },

>     state:          { type: "string", enum: ["Texas", "Alabama", "California", "other"] },

Reviewer 4: ```suggestion
```

  },
  required: ["street_name", "city", "state", "house_number", "state"],
};

const interfaceModels = await generator.generate(schema);

> // generated interfaceModels[0] should have the following shape:
> export interface Address {
>   streetName: string;
>   // City description
>   city: string;
>   state: string;
>   houseNumber: number;
>   // Status if marriage live in given house
>   marriage?: boolean;
>   petNames?: Array<string>;
>   state?: "Texas" | "Alabama" | "California" | "other";
> }

Reviewer 4: ```suggestion
// generated interfaceModels[0].result should have the following shape:
export interface Address {
  streetName: string;
  // City description
  city: string;
  houseNumber: number;
  // Status if marriage live in given house
  marriage?: boolean;
  petNames?: Array<string>;
}
```

```

Some important resources to follow:
- [How processing data from input works](https://github.com/asyncapi/generator-model-sdk/blob/master/docs/input_processing.md)
- [How **generator-model-sdk** simplify JSON Schema](https://github.com/asyncapi/generator-model-sdk/blob/master/docs/simplification.md)
- [How model generator works and how to write a new one](https://github.com/asyncapi/generator-model-sdk/blob/master/docs/generators.md)
- [How presets (customization) works and how to write a custom preset](https://github.com/asyncapi/generator-model-sdk/blob/master/docs/customization.md)

## Community continues to grow

In February we reached 1600 followers on Twitter.

<TwitterTweetEmbed
  tweetId='1362016089207635968'
  options={{
    cards: 'hidden',
    width: 500,
    align: 'center'
  }}
/>

A great thing to see is that the community from [LinkedIn](https://www.linkedin.com/feed/update/urn:li:activity:6769647615823945728/) grew to 900, and our news shared there also get more visibility.

We are very noisy :sweat_smile:

## Other features

> - Thanks to the support of [Daniel CHU](https://www.linkedin.com/in/daniel-chu-dc/) that joined us during last Hacktoberfest now the JavaScript parser also validates the examples of server variables. For more details, check [1.4](https://github.com/asyncapi/parser-js/releases/tag/v1.4.0) release.

Reviewer 2: ```suggestion
- Thanks to the support of [Daniel CHU](https://www.linkedin.com/in/daniel-chu-dc/) —who joined us during last Hacktoberfest— now the JavaScript parser also validates the examples of server variables. For more details, check out the [1.4](https://github.com/asyncapi/parser-js/releases/tag/v1.4.0) release.
```

> - Once again, thanks to [Ludovic Dussart](https://twitter.com/ldussart) from Ineat Lab, we have additional features, these time in HTML and Markdown template. Thanks to the new **version** parameter, you can overwrite the version of the application specified in the AsyncAPI file under **info.version**. Useful in CI/CD when your service version is not maintained in the #AsyncAPI file but pom.xml or package.json

Reviewer 2: ```suggestion
- Once again, thanks to [Ludovic Dussart](https://twitter.com/ldussart) from Ineat Lab, we have additional features, these time in HTML and Markdown template. Thanks to the new **version** parameter, you can overwrite the version of the application specified in the AsyncAPI file under **info.version**. Useful in CI/CD when your service version is not maintained in the AsyncAPI file but pom.xml or package.json
```

> - Thanks to [Mike Ralphson](https://twitter.com/PermittedSoc), we have a new **frontMatter** parameter in the Markdown template. It lets you specify an external file with a custom front-matter that should be included in generated Markdown during generation. Very useful for static site generators' users.

Reviewer 2: ```suggestion
- Thanks to [Mike Ralphson](https://twitter.com/PermittedSoc), we have a new **frontMatter** parameter in the Markdown template. It lets you specify an external file with a custom front-matter that should be included in the resulting Markdown during generation. Very useful for static site generators' users.
```

  ```bash
  # 1. Get AsyncAPI Generator
  npm install -g @asyncapi/generator

  # 2. Create a file with frontmatter
  cat > ssg.yml << EOF
  title: AsyncAPI Documentation
  layout: asyncapi
  permalink: /asyncapi-docs
  EOF

  # 3, Generate Markdown file that includes the frontmatter
  ag https://bit.ly/asyncapi @asyncapi/markdown-template -o output -p frontMatter=ssg.yml

  # 4. Check out the output
  cat output/asyncapi.md
  ```
- [Maciej Urbanczyk](https://www.linkedin.com/in/maciej-urba%C5%84czyk-909547164/) and [I](https://twitter.com/derberq) (I mean me :smiley:) pushed some features to the AsyncAPI Generator:
  - It supports the latest Node.js 15 and npm 7
  - You can now install generator templates globally. For more details, read [this](https://github.com/asyncapi/generator/#global-templates-installed-with-yarn-or-npm) new section in the readme.

>   - It is now much easier to generate multiple files using the new React render engine

Reviewer 4: Please add reference to docs -> https://github.com/asyncapi/generator/blob/master/docs/authoring.md#react and for example in real template: https://github.com/asyncapi/template-for-generator-templates/blob/master/template/schemas/schema.js#L10

>   - You can now use TypeScript in all parts of your generator templates

Reviewer 4: Users can also use from `1.4.0` generator asynchronous filters in Nunjucks - please refer to the docs https://github.com/asyncapi/generator/blob/master/docs/authoring.md#filters

Reviewer 4: Also this is not a true, because we have still problem to use TS in React templates - I have problem in my branch to transpile TS to JS (problem with imports etc). TS we can only use at the moment in hooks and Nunjucks filters, not in `/template` folder using React render engine 

Please add reference to the docs for TS -> https://github.com/asyncapi/generator/blob/master/docs/authoring.md#typescript-support

Author: I did not mention async filters Nunjucks on purpose. I do not want to promote Nunjucks in my blog posts. 

for the TS issue, is there a follow-up issue?

Reviewer 4: Like in this meme:

> Nunjucks is not my friend anymore. My new friend is React XD. 

You should remove blog posts related to Nunjucks (I'm just kidding 😄).

Yes, there is an issue: https://github.com/asyncapi/generator-react-sdk/issues/3

Author: ok, I changed text a bit

Reviewer 4: Also, please have in mind that without asynchronous filters you cannot use `generator-model-sdk` in Nunjucks, so maybe we should mention about Nunjucks last time in blog post?

Author: nah, we do not even support model-sdk in one common way in the generator, it is a detail not worth mentioning in simple status update

Reviewer 4: So ok then.

Check out all the [releases](https://github.com/asyncapi/generator/releases)

## Good learning resources

- Read this important [10 FAQs About Event-Driven APIs](https://duckster.medium.com/10-faqs-about-event-driven-apis-50ee9c94bbb8) from [Dakshitha Ratnayake](https://twitter.com/techieducky)
- Great summary from Nordic APIs and [Vyom Srivastava](https://www.linkedin.com/in/vyomsrivastava/) on [8+ AsyncAPI Documentation Generators](https://nordicapis.com/8-asyncapi-documentation-generators/)

> - Fran Mendez was a part of Postman's live stream. Learn how to get from zero to AsyncAPI in just about 1h 30min. I hope it can be done faster, and Fran was just having too much small talk with Kevin and Kin :sweat_smile:

Reviewer 2: 😂
